### PR TITLE
VTK output improvements

### PR DESCRIPTION
**Pull request by klevzoff**
This PR improves a few aspects of VTK output:

1. Adds the ability to associate component labels with inner dimensions of N-D arrays. Labels are set by physics modules (solvers or constitutive models) that own arrays, stored in the `Wrapper` and used by VTK output (can also use them in Silo or Blueprint in the future). This makes for a really nice presentation in ParaView, see below for examples. This led to deleting the specialization for 4D arrays added in https://github.com/GEOSX/GEOSX/pull/1464 (sorry @francoishamon ... but your PR is what motivated me to look deeper into VTK capabilities!) and also streamlining some of the code to handle arbitrary dimensions.

2. Adds averaging over quadrature points for material data. I took advantage of the same averaging machinery used by @corbett5 in BlueprintOutput which worked very well. This drastically reduced the amount of data written to disk.

3. Instructs VTK to use proper native type for each array. Previously all data, including integral arrays such as ghost rank, were forced into arrays of `double`... not anymore. This also saves disk space/bandwidth by a tiny fraction. Perhaps there are also benefits to ParaView recognizing integral array types, though I haven't seen one yet.

4. Removes support for arrays of `R1Tensor`. We haven't been using them for quite some time, so I took the liberty of removing them entirely from `rtTypes`, lambda dispatch and outputs drivers. This will also discourage developers from using it in future code. In VTK this leads to code simplification, in particular, no more need for custom array type `VTKGEOSXData` which is removed. There are still traces of it in Silo, but it's not worth cleaning if we're going to deprecate Silo at some point.

5. Miscellaneous cleanups in VTK code and elsewhere (apologies to reviewers, I have troubles keeping PR laser focused).

Passes unit/integrated tests, no rebaseline needed, ready to merge upon review.

---
Examples of using component labels:

Solid stress (Voight tensor components labeled):
stress

Fluid 1-d property (phase density):
phaseDensity

Fluid 2-d property (phase compositions; we can change how compound labels are formed, for now I concatenate them with `/`):
phaseCompFrac

(I don't know which ParaView version is the earliest to support arbitrary number of components, but it worked for me with 5.9.1)

### CODE REVIEW

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp [hunk 1]
```diff
@@ -296,230 +246,299 @@ VTKPolyDataWriterInterface::getVtkCells( CellElementRegion const & er ) const
...
+  subRegions.forSubGroups< SUBREGION >( [&]( SUBREGION const & esr )
+  {
+    numElements += esr.size();
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    if( first )
     {
-      writeElementField4D< SUBREGION >( celldata, er, field );
+      rtTypes::applyArrayTypeLambda2( rtTypes::typeID( wrapper.getTypeId() ), true, [&]( auto array, auto scalar )
+      {
+        data = vtkAOSDataArrayTemplate< decltype( scalar ) >::New();
```
Comment (klevzoff) [inline src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:428]:
```suggestion
```

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp [hunk 2]
```diff
@@ -296,230 +246,299 @@ VTKPolyDataWriterInterface::getVtkCells( CellElementRegion const & er ) const
...
+        // with a known material name. I can't think of an easier way at the moment, but there ought to be one.
+        auto const startsWithMaterial = [&]( string const & mat ){ return name.substr( 0, mat.size() ) == mat; };
+        if( std::find_if( er.getMaterialList().begin(), er.getMaterialList().end(), startsWithMaterial ) != er.getMaterialList().end() )
+        {
+          materialFields.insert( name );
+        }
+        else
+        {
+          regularFields.insert( name );
+        }
```
Comment (klevzoff) [inline src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:509]: Ideas?

**Review by TotoGaz — COMMENTED** [review]
I would require mentoring to go deeper in the pure `vtk` stuff.

File: src/coreComponents/common/Span.hpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+/**
+ * @brief Lightweight non-owning wrapper over a contiguous range of elements.
+ * @tparam T type of range element
+ *
+ * This is a simplified version of C++17's std::span<> that doesn't support
+ * compile-time extents.
+ */
+template< typename T >
+class Span
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/common/Span.hpp:39]: Yes, `gsl::span` came to mind, but to be honest, writing a simplified implementation myself probably took me less time than: a) deciding which implementation to include, b) deciding on submodule vs TPL route, c) making sure there are no compilation issues or conflicts (for example, their page clams Clang 10 and 11 supported, but we're building with Clang 8). I don't like reinventing the bicycle as much as anyone, but this is one case where I think it's fine (of course, the caveat is I haven't written any test for it, which will take way more time than the class itself). I don't have strong feelings about it though, I'll happily replace it with something else if everyone agrees it's a better way.

As for `std::span`, it should be more or less a straightforward find/replace, because I kept mine more or less a subset of `std::span`'s interface (albeit, c++20 version), with fewer constructors and lacking a few methods and a defaulted template parameter. Speaking of which, I'd like us to have a conversation about moving to C++17 soon :)

[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/common/Span.hpp:39]: That's also the opportunity to discuss how we could do this kind of inclusions for other cases that are header only.
For example, I was using https://github.com/nlohmann/json. Just
```cmake
include(FetchContent)

FetchContent_Declare(json
  GIT_REPOSITORY https://github.com/nlohmann/json.git
  GIT_TAG v3.7.3)

FetchContent_GetProperties(json)
if(NOT json_POPULATED)
  FetchContent_Populate(json)
  add_subdirectory(${json_SOURCE_DIR} ${json_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

target_link_libraries(foo PRIVATE nlohmann_json::nlohmann_json)
```
and you're done, that's cool.

> Speaking of which, I'd like us to have a conversation about moving to C++17 soon :)

Save your breath.

C++ 17 will surely [require CUDA 11+](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#c-cplusplus-language-support) which on [Power requires RHEL 8](https://developer.nvidia.com/cuda-11.0-download-archive?target_os=Linux&target_arch=ppc64le&target_distro=RHEL&target_version=8) where Pangea3 is RHEL 7.
And I have no insight about any update.
Furthermore, CUDA 10 does not accept g++ versions greater than 8 (if I remember correctly), so we cannot even try to trick some features in... (which may be better 🤣 )

And that's so sad to be kept with old C++ 😢

File: src/coreComponents/codingUtilities/StringUtilities.hpp [hunk 4]
```diff
@@ -60,6 +60,16 @@ string join( IT first, IT last, S const & delim = S())
   return oss.str();
 }
 
+template< typename S = char, typename T, typename ... Ts >
+string concat( S const & delim, T const & v, Ts const & ... vs )
+{
+  std::ostringstream oss;
+  oss << v;
+  using expander = int[];
+  (void) expander{ 0, ( void (oss << delim << vs ), 0 )... };
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/codingUtilities/StringUtilities.hpp:69]: Ah yes, I lifted this "fold expression" technique from LvArray:
https://github.com/GEOSX/LvArray/blob/b80e1b1df9ad2d3e9163d7a5ff088f0e3895ebb8/src/indexing.hpp#L141-L152
(I couldn't use that function directly because it enforces a slightly different format than what I needed).

The leading 0 is needed to account for the case where the variadic pack is empty (0-element arrays are not allowed). This case is currently not triggered anywhere, but it could be in the future.

The `void` cast is indeed to silence warnings, and it seems like it cannot be well-formed without a separate `using` declaration.

File: src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp [hunk 5]
```diff
@@ -104,12 +107,12 @@ class CapillaryPressureBase : public ConstitutiveBase
   CapillaryPressureBase( string const & name,
                          dataRepository::Group * const parent );
 
-  virtual ~CapillaryPressureBase() override;
+  virtual ~CapillaryPressureBase() override = default;
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp:110]: Yes, and I've seen your comments on this in other PRs. While I agree (and so does CLion, with very in-my-face highlighting) that virtual is not needed, it is the current code style and I keep it everywhere for consistency. IIRC, it came up in a discussion once or twice, and the argument for keeping `virtual` was it makes virtual functions more obvious in a class (due to being the first thing in a function decl). If you want to make an argument for removing it repository-wide, I will support it :)

File: src/coreComponents/constitutive/fluid/MultiFluidBase.cpp [hunk 6]
```diff
@@ -105,38 +105,60 @@ MultiFluidBase::MultiFluidBase( string const & name, Group * const parent )
...
   m_dTotalDensity_dPressure.resize( size, numPts );
   m_dTotalDensity_dTemperature.resize( size, numPts );
-  m_dTotalDensity_dGlobalCompFraction.resize( size, numPts, NC );
+  m_dTotalDensity_dGlobalCompFraction.resize( size, numPts, numComp );
+}
+
+void MultiFluidBase::setLabels()
+{
+  Span< string const > const compNames( m_componentNames.begin(), m_componentNames.end() );
+  Span< string const > const phaseNames( m_phaseNames.begin(), m_phaseNames.end() );
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:145]: Yes, good point, that's the (explicitly) correct thing to do.

File: src/coreComponents/codingUtilities/StringUtilities.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/codingUtilities/StringUtilities.hpp:69]: > The leading 0 is needed to account for the case where the variadic pack is empty (0-element arrays are not allowed). This case is currently not triggered anywhere, but it could be in the future.

That's what I was thinking and I actually tried on godbolt.org with multiple compilers.
Getting rid of the leading `0` works smoothly.
So I was wondering if I was missing something. 🤷 

My guess would be that the initial _stackoverflow_ answer gets the extra zero before it did invade the world! 🤣 🚀

File: src/coreComponents/codingUtilities/StringUtilities.hpp [hunk 7]
```diff
@@ -60,6 +60,16 @@ string join( IT first, IT last, S const & delim = S())
   return oss.str();
 }
 
+template< typename S = char, typename T, typename ... Ts >
+string concat( S const & delim, T const & v, Ts const & ... vs )
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/codingUtilities/StringUtilities.hpp:64]: Yes! Don't know how doxygen let me off this one.

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp (hunk 2 above)
Comment (TotoGaz) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:509]: Hmm, I'm afraid I wrote something wrong, be careful with my suggestion.

File: src/coreComponents/codingUtilities/StringUtilities.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/codingUtilities/StringUtilities.hpp:69]: It works by default, but not with `-pedantic`: https://godbolt.org/z/hsxnzrrPs
Also, the standard says "N shall be greater than zero": http://eel.is/c++draft/dcl.array#1
So I'll keep it to be on the safe side :safety_vest:

File: src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.cpp [hunk 8]
```diff
@@ -26,22 +26,31 @@ using namespace dataRepository;
...
+  {
+    { "gas", CapillaryPressureBase::PhaseType::GAS },
+    { "oil", CapillaryPressureBase::PhaseType::OIL },
+    { "water", CapillaryPressureBase::PhaseType::WATER }
+  };
+  auto const it = phaseDict.find( lookup );
+  if( it == phaseDict.end() )
+  {
+    std::vector< string > phaseNames;
+    std::transform( phaseDict.begin(), phaseDict.end(), std::back_inserter( phaseNames ), []( auto const & p ){ return p.first; } );
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.cpp:44]: Yeah, performance is not a concern (and is equivalent in this case). Expressiveness is actually my main goal when I try to use standard algorithms. Unfortunately, current C++ and classic STL make it a little too verbose with iterators. 

I'd argue that with ranges (C++20), this
```c++
auto phaseNames = phaseDict | std::views::transform( []( auto && p ){ return p.first; } );
```
spells out "extract keys from the dict" much cleaner than the loop version, and C++ devs should get used to more functional style of programming :smile: 
(this also has performance benefits, as it constructs a lazily evaluated range that you can pass directly to `join` or `std::format` or some other function that concatenates the values, and avoid allocating the vector).

But back to reality... yeah, I guess I don't mind either way.

File: src/coreComponents/dataRepository/WrapperBase.hpp [hunk 9]
```diff
@@ -548,11 +550,67 @@ class WrapperBase
...
+   * @brief Downcast @p this to a typed wrapper.
+   * @tparam T wrapped type to cast to
+   * @pre @p this must be in instance of Wrapper<T>
+   * @return reference to @p this cast to a typed wrapper.
+   */
+  template< typename T >
+  Wrapper< T > & cast()
+  {
+    GEOSX_ERROR_IF( getTypeId() != typeid( T ), "WrapperBase: invalid downcast" );
+    return dynamicCast< Wrapper< T > & >( *this );
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/dataRepository/WrapperBase.hpp:562]: What I want to do (and am using extensively in this PR) is
```c++
WrapperBase & wrapper = ... // get the wrapper from the group
dispatch( wrapper->getTypeId(), [&]( auto value )
{
  using T = decltype( value );
  auto & wrapperT = wrapper.cast< T >; // previously, this would've been a longer expression
  // use the wrapperT to access the stored value
} );
```

I don't think there is a way to do this without forward declaring `Wrapper< T >` in `WrapperBase.hpp`. Sure, it seems like a circular dependency, but it isn't, and there is no include cycle (the user ultimately just includes `Wrapper.hpp` once which pulls in `WrapperBase.hpp`, and that's it).

Sure, I could force the user to type
```c++
auto & wrapperT = wrapper.cast< Wrapper< T > >; 
```
but that seems like boilerplate, because `Wrapper` is the only thing that inherits from `WrapperBase`, so no other type could possibly be the target of the cast.

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp [hunk 10]
```diff
@@ -296,230 +246,299 @@ VTKPolyDataWriterInterface::getVtkCells( CellElementRegion const & er ) const
...
+/**
+ * @brief Checks consistency of per-dimension labels (must either be not present, or match the size of array).
+ * @param wrapper the array wrapper
+ * @param dim dimension index to check
+ */
+template< typename T, int NDIM, typename PERM >
+void checkLabels( Wrapper< Array< T, NDIM, PERM > > const & wrapper, int const dim )
+{
+  Span< string const > const labels = wrapper.getDimLabels( dim );
+  if( !labels.empty() )
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:303]: I'm currently calling `checkLabels()` outside of `!labels.empty()` check in `getDimLabels`. I could move the call inside the branch and then indeed wouldn't need to handle empty case here. I would have to change ternary operator in `getDimLabels` into an `if/else` statement though, but I guess that's hardly a reason not to have a more logical control flow.

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp (hunk 2 above)
Comment (klevzoff) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:509]: I guess my question was indeed more on the content side: do we have a good way in our data repository structure to differentiate between "normal" fields allocated by solvers and "material" fields allocated by constitutive models? The latter have an extra gauss point dimension that needs to be handled separately. I'm doing that based on field name prefixes, but that seems super brittle. The user could name a material something that happens to be a prefix of an existing field name in one of the solvers.

File: src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp [hunk 11]
```diff
@@ -92,10 +92,13 @@ class CapillaryPressureBase : public ConstitutiveBase
...
-    static constexpr integer OIL            = 0;
-    static constexpr integer GAS            = 1;
-    static constexpr integer WATER          = 2;
-    static constexpr integer MAX_NUM_PHASES = 3;
+    enum : integer
+    {
+      OIL            = 0,
+      GAS            = 1,
+      WATER          = 2,
+      MAX_NUM_PHASES = 3,
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp:100]: Yeah. I just replaced `static constexpr` variables with unscoped enums because the latter don't require a definition in the cpp file (although I was the one who recommended using `static constexpr` in the first place...). Although we (currently) don't switch on these values, I agree that `MAX_NUM_PHASES` should perhaps be declared outside of enum.

File: src/coreComponents/codingUtilities/StringUtilities.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/codingUtilities/StringUtilities.hpp]: OK, this is precise!

File: src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.cpp:44]: Yes, `ranges` seems to be great.

I would argue that it's maybe an element that would bring a big big booster to GEOSX codebase.
We use extensively this kind of methods
```c++
class Group {
...
  template< typename GROUPTYPE = Group, typename ... GROUPTYPES, typename LAMBDA >
  void forSubGroups( LAMBDA && lambda ) const
...
}
```
With a predicate (only based on type by design) and with a callback/functor pattern, at the noteworthy difference w.r.t the STL that the dev using these functions does not own the container on which it operates. Making the chaining of these calls more complicated. We sometimes end in nested lambda and nested captures... 

Also, in some cas... Oh, I realize I was going to repeat https://github.com/GEOSX/GEOSX/issues/1089!

There exists a c++14 implementation of the `ranges` lib https://github.com/ericniebler/range-v3.
_I'm convinced that this could literally change GEOSX for the best._
So we may want to include it. That would be a smart move imho.

> C++ devs should get used to more functional style of programming 😄

This is actually not *only* a functional pattern issue.
It's just that the boilerplate that C++ brings is not worth the price here 😉 

Let's assume that you want to filter a list and apply a function on it.
In `python` you have
```python
map( f, filter( p, lst ) )
```
In `java` (and maybe `javascript` with the same syntax)
```java
stream.filter( p ).map( f )
```
in `common lisp` something like
```lisp
( mapcar f ( remove-if-not p lst ) )
```
You are young, you're surely brave enough to tell me how to do this in std C++ without `ranges`!

File: src/coreComponents/dataRepository/WrapperBase.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/dataRepository/WrapperBase.hpp:562]: > Sure, it seems like a circular dependency, but it isn't

We are here in the case where `Wrapper` depends on `WrapperBase` with inheritance and `WrapperBase` depends on `Wrapper` though `WrapperBase::cast`.
You manage to get it work because of the flexibility of C++ on this specific point.
This is a cycle and I think that in a language like `java` this would not even compile.
Moreover, having a base class know about its derived class really has the smell of wrong design here.

> there is no include cycle (the user ultimately just includes Wrapper.hpp once which pulls in WrapperBase.hpp, and that's it).

There is no include cycle because you forward declare. Otherwise, 
- To know about `WrapperBase`, `Wrapper.hpp` needs to include `WrapperBase.hpp`.
- To know about `Wrapper`, `WrapperBase.hpp` needs to include `Wrapper.hpp`.

That the developer just includes `Wrapper.hpp` is OK, but the semantic cycle is behind the scenes.

Now, could it be possible to have a nested static function (or simple free function `wrapperCast`) inside of `Wrapper` that would cast for you?
Calling `Wrapper< T >::cast( wrapperBase )` instead of `wrapperBase.cast< T >()` would do the same (at a boilerplate cost that seems reasonable).
What do you think of this option?

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp]: It's hard to follow the flow on github, I would have to pull the code for that.
It's just that if `labels` is empty, then `wrapper.reference().size( dim )` should be zero to be uniform with other cases. What is the purpose of your `getDefaultLabels` then?

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp (hunk 2 above)
Comment (TotoGaz) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:505]: Yes, I understood that it was about content. But I think that my code snippet is wrong nevertheless. 😉

File: src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp]: No we do not `switch` on `enums`, but maybe that could be a refactoring idea in the future if we manage to find some resources more dedicated to refactoring and pure code  🚀

File: src/coreComponents/common/Span.hpp [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+public:
+
+  /// Type of range element
+  using element_type = T;
+
+  /// Type of underlying value
+  using value_type = std::remove_cv_t< T >;
+
+  /// Type used for indexing the range
+  using size_type = std::ptrdiff_t;
```
Comment (TotoGaz) [inline src/coreComponents/common/Span.hpp:51]: Actually `size_type` should preferably be a `std::size_t`, shouldn't it?

Comment (klevzoff) [reply src/coreComponents/common/Span.hpp:51]: I made it signed intentionally, because our Array types use `localIndex` (which is a signed type) as size type, and this just makes things like `labels.size() == array.size( dim )` easier to write.

In general, there seems to be an increasing understanding among c++ committee people that unsigned size and indexing were probably a mistake. It's the reason why [std::ssize](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1227r1.html) was introduced. There is also some rationale given here: https://github.com/ericniebler/stl2/issues/182#issuecomment-257000171

If we want to strictly stick to `std::span`, I should make this `std::size_t`... and add static casts everywhere I use it. Do we want this, is the question?

Comment (TotoGaz) [reply src/coreComponents/common/Span.hpp:51]: I would preferably stick to the standard so we can replace `geosx::Span` by the `std` version with less risks.

And since the `std` solution seems to use a `std::ssize`, do you think adding a simple implementation in GEOSX (like provided in https://en.cppreference.com/w/cpp/iterator/size) could provide the most standard solution?
When we move to C++20, this can be replaced by `std` (we can add a `// LEGACY_C++20` comment inside the function to find them!).

Instead of having casts here and there in the code, would it make sense to replace them by `ssize`?

File: src/coreComponents/dataRepository/WrapperBase.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rrsettgast) [reply src/coreComponents/dataRepository/WrapperBase.hpp:562]: This used to be in `Wrapper.hpp` but someone took it out because it wasn't being used at the time.

```
  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted Wrapper<T>
   */
  static Wrapper< T > * cast( WrapperBase * const base )
  {
#ifdef USE_DYNAMIC_CASTING
    return dynamic_cast< Wrapper< T > * >(base);
#else
    return static_cast< Wrapper< T > * >(base);
#endif
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted reference to const Wrapper<T>
   */
  static Wrapper< T > const * cast( WrapperBase const * const base )
  {
#ifdef USE_DYNAMIC_CASTING
    return dynamic_cast< Wrapper< T > const * >(base);
#else
    return static_cast< Wrapper< T > const * >(base);
#endif
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted Wrapper<T>
   */
  static Wrapper< T > & cast( WrapperBase & base )
  {
    if( base.get_typeid() != typeid(T) )
    {
      GEOS_ERROR( "invalid cast attempt" );
    }
    return static_cast< Wrapper< T > & >(base);
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted reference to const Wrapper<T>
   */
  static Wrapper< T > const & cast( WrapperBase const & base )
  {
    if( base.get_typeid() != typeid(T) )
    {
      GEOS_ERROR( "invalid cast attempt" );
    }
    return static_cast< Wrapper< T > const & >(base);
  }
```

[the remark this answers is not in the document]
Comment (corbett5) [reply src/coreComponents/dataRepository/WrapperBase.hpp:562]: Pretty sure that someone was me.

File: src/coreComponents/dataRepository/wrapperHelpers.hpp [hunk 13]
```diff
@@ -98,7 +99,47 @@ real64 const * getPointerToComponent( R1Tensor const & var, int const component
...
+{
+public:
+
+  void set( integer const dim, Span< string const > labels )
+  {
+    GEOSX_ERROR_IF_LT( dim, 0 );
+    GEOSX_ERROR_IF_GE( dim, NDIM );
+    m_values[dim].resize( labels.size() );
+    std::copy( labels.begin(), labels.end(), m_values[dim].begin() );
+  }
```
Comment (rrsettgast) [inline src/coreComponents/dataRepository/wrapperHelpers.hpp:130]: Maybe a check to make sure that the the size of the `dim` dimension is equal to the number of labels?

I know this is a chicken or egg situation. You are setting the labels prior to the resize of the actual array, but then the resize should check to make sure it has the appropriate resize for a dimension corresponding to the number of labels in `m_dimLabels.m_values`.  Maybe this is already done and I missed it.

File: src/coreComponents/dataRepository/wrapperHelpers.hpp [hunk 14]
```diff
@@ -627,20 +668,29 @@ std::unique_ptr< int > averageOverSecondDim( T const & )
   return std::unique_ptr< int >( nullptr );
 }
 
-template< typename T >
-typename std::enable_if_t< traits::is_array_type< T >, int >
-numArrayDims( T const & GEOSX_UNUSED_PARAM( var ) )
+template< typename T, int NDIM, int USD >
+int numArrayDims( ArrayView< T const, NDIM, USD > const & GEOSX_UNUSED_PARAM( var ) )
```
Comment (rrsettgast) [inline src/coreComponents/dataRepository/wrapperHelpers.hpp:672]: Can't we just use `var.NDIM` or put a member function in `ArrayView` to give access to NDIM?

**Review by rrsettgast — APPROVED** [review]
Just a minor objection about the cast in WrapperBase. Resolve or move it and you are good IMO.

File: src/coreComponents/dataRepository/wrapperHelpers.hpp (hunk 13 above)
Comment (klevzoff) [reply src/coreComponents/dataRepository/wrapperHelpers.hpp:130]: Yeah, that's exactly the doubt I've had. Since resizing the array does _not_ happen through `Wrapper` but instead through a direct reference, there is no way to always enforce the invariant. So I decided to check "eventual consistency" only at usage, i.e. in output code. 

Adding the check here can prevent _some_ potential bugs, but it will also prevent the following usage (actually seen in code):
```c++
subRegion.getWrapper< array2d< real64, compflow::LAYOUT_COMP > >( viewKeyStruct::globalCompDensityString() ).
      setDimLabels( 1, fluid.componentNames() ).
      reference().resizeDimension< 1 >( m_numComponents );
```
Here I cannot reorder the calls because once you get `reference()`, wrapper method chaining is lost. I'll have to unchain it and store reference to wrapper in a local variable, or just move setting labels to a different function entirely.

File: src/coreComponents/dataRepository/wrapperHelpers.hpp (hunk 14 above)
Comment (klevzoff) [reply src/coreComponents/dataRepository/wrapperHelpers.hpp:672]: That's what the previous version did, returning `T::NDIM`. I think this way is cleaner than using `traits::is_array_type` and `enable_if`.

File: src/coreComponents/common/Span.hpp (hunk 12 above)
Comment (klevzoff) [reply src/coreComponents/common/Span.hpp:51]: Fair enough, I'll change to `std::size_t`

File: src/coreComponents/dataRepository/WrapperBase.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/dataRepository/WrapperBase.hpp:562]: Sounds good, I will place the (reference) downcast in `Wrapper< T>`, if there are no objections

File: src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp (hunk 2 above)
Comment (rrsettgast) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:506]: We used to have a differentiation between regular fields and material fields....but folks weren't happy about having to reach into an extra layer for material data access, so I flattened it. I am not opposed to adding the layer back in.

Comment (klevzoff) [reply src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:506]: I ended up collecting material fields by iterating over constitutive models directly rather than proxy wrappers on subregions, then filtering out those names from the list of subregion wrappers to get the rest. In other words, VTK output will now also be using the extra layer of access.

I'm not sure there are any pieces of code left that use the "flattened" data model. Definitely not any physics solvers I'm aware of, nor Silo or Blueprint. I think if someone wanted to collect a time history of a material field (say stress), they would have to specify `fieldName="rock_stress"` in a `PackCollection` task and it would collect through this proxy wrapper as there is no other way currently. Other than than, we could remove the material data wrappers from subregions.
